File: julia/test/test_snap.jl
module TestSnap
using Test
import GraphIdx.Io.Snap:
  parse_snap_idx, parse_snap_header, parse_snap_edges
import GraphIdx


const TEST_SNAP = """# comments
# ... to be ignored
#
# Nodes: 4 Edges: 3
# FromNodeId    ToNodeId
1       2
1       4
3       4
"""

@testset "SNAP reader                    " begin
    idx = parse_snap_idx(IOBuffer(TEST_SNAP))
    n, m, head, tail = parse_snap_edges(IOBuffer(TEST_SNAP), Int)
    @test parse_snap_header(IOBuffer(TEST_SNAP)) == (4, 3)
    @test GraphIdx.num_nodes(idx) == 4
    @test GraphIdx.num_edges(idx) == 3
    @test n == 4
    @test m == 3
    @test head == [1, 1, 3]
    @test tail == [2, 4, 4]
    @test idx == GraphIdx.BiAdjacentIndex([0, 2, 3, 4, 6] .+ 1,
                                          [1, 3, 0, 3, 0, 2] .+ 1)
end


end

File: cxx/test/test_snap.cpp
#include <doctest/doctest.h>
#include <sstream>
#include "../graphidx/io/snap.hpp"


const auto GRAPH = R"(# comments
# ... to be ignored
#
# Nodes: 4 Edges: 3
# FromNodeId    ToNodeId
1       2
1       4
3       4
)";


TEST_CASE("snap: header")
{
    std::istringstream io (GRAPH);
    size_t n = 0, m = 0;
    parse_snap_header(io, n, m);
    REQUIRE(n == 4);
    REQUIRE(m == 3);
    std::string line;
    std::getline(io, line);
    REQUIRE(line == "1       2");
}


TEST_CASE("snap: biadj")
{
    std::istringstream io (GRAPH);
    auto idx = parse_snap_idx(io);
    CHECK(std::vector<int>(idx.value) == std::vector<int> {1, 3, 0, 3, 0, 2});
    CHECK(idx.index == std::vector<int> {0, 2, 3, 4, 6});
    REQUIRE(idx == BiAdjacent {{1, 3, 0, 3, 0, 2},
                               {0, 2, 3, 4, 6}});
}
